src/main/java/teammates/logic/api/EmailGenerator.java

```diff
@@ -1,2 +1,1083 @@
+import teammates.common.datatransfer.attributes.StudentAttributes;
+import teammates.common.util.Config;
+import teammates.common.util.Const;
+import teammates.common.util.EmailType;
+import teammates.common.util.EmailWrapper;
+import teammates.common.util.RequestTracer;
+import teammates.common.util.SanitizationHelper;
+import teammates.common.util.Templates;
+import teammates.common.util.Templates.EmailTemplates;
+import teammates.common.util.TimeHelper;
+import teammates.logic.core.CoursesLogic;
+import teammates.logic.core.FeedbackSessionsLogic;
+import teammates.logic.core.InstructorsLogic;
+import teammates.logic.core.StudentsLogic;
+
+/**
+ * Handles operations related to generating emails to be sent from provided templates.
+ *
+ * @see EmailTemplates
+ * @see EmailType
+ * @see EmailWrapper
+ */
+public final class EmailGenerator {
+    // feedback action strings
+    private static final String FEEDBACK_ACTION_SUBMIT_EDIT_OR_VIEW = "submit, edit or view";
+    private static final String FEEDBACK_ACTION_VIEW = "view";
+    private static final String FEEDBACK_ACTION_SUBMIT_OR_UPDATE =
+                            ", in case you have not submitted yet or wish to update your submission. ";
+    private static final String HTML_NO_ACTION_REQUIRED = "<mark>No action is required if you have already submitted</mark>";
+
+    // status-related strings
+    private static final String FEEDBACK_STATUS_SESSION_OPEN = "is still open for submissions"
+                                        + FEEDBACK_ACTION_SUBMIT_OR_UPDATE + HTML_NO_ACTION_REQUIRED;
+    private static final String FEEDBACK_STATUS_SESSION_OPENING = "is now open";
+    private static final String FEEDBACK_STATUS_SESSION_CLOSING = "is closing soon"
+                                        + FEEDBACK_ACTION_SUBMIT_OR_UPDATE + HTML_NO_ACTION_REQUIRED;
+    private static final String FEEDBACK_STATUS_SESSION_CLOSED = "is now closed for submission";
+    private static final String FEEDBACK_STATUS_SESSION_OPENING_SOON = "is due to open soon";
+
+    private static final String DATETIME_DISPLAY_FORMAT = "EEE, dd MMM yyyy, hh:mm a z";
+
+    private static final long SESSION_LINK_RECOVERY_DURATION_IN_DAYS = 90;
+
+    private static final EmailGenerator instance = new EmailGenerator();
+
+    private final CoursesLogic coursesLogic = CoursesLogic.inst();
+    private final FeedbackSessionsLogic fsLogic = FeedbackSessionsLogic.inst();
+    private final InstructorsLogic instructorsLogic = InstructorsLogic.inst();
+    private final StudentsLogic studentsLogic = StudentsLogic.inst();
+
+    private EmailGenerator() {
+        // prevent initialization
+    }
+
+    public static EmailGenerator inst() {
+        return instance;
+    }
+
+    /**
+     * Generates the feedback session opening emails for the given {@code session}.
+     */
+    public List<EmailWrapper> generateFeedbackSessionOpeningEmails(FeedbackSessionAttributes session) {
```

Comment:
Consider wrapping this in with the generateFeedbackSessionClosingEmails method as they are very similar. A one line boolean would do the trick.